go-controller/pkg/factory/factory.go

```diff
@@ -26,16 +26,23 @@ type informer struct {
 }
 
 func (i *informer) forEachHandler(obj interface{}, f func(id uint64, handler cache.ResourceEventHandler)) {
+	handlersCopy := make(map[uint64]cache.ResourceEventHandler)
+
 	i.Lock()
-	defer i.Unlock()
+	// copying the informer handlers and releasing the lock prevents deadlocks when trying to change
+	// handlers within a handler
+	for key, value := range i.handlers {
+		handlersCopy[key] = value
+	}
+	i.Unlock()
 
 	objType := reflect.TypeOf(obj)
 	if objType != i.oType {
 		logrus.Errorf("object type %v did not match expected %v", objType, i.oType)
 		return
 	}
 
-	for id, handler := range i.handlers {
+	for id, handler := range handlersCopy {
```

Comment:
This change is not good.
Because we might now be calling a handler that is now expected to be deleted. 
1. We copied the handler and are operating on it.
2. The lock is released meanwhile and it may be used to delete the handler from the map (i.e. we are not expecting to receive any events to be delivered)
3. The operation from step #1 will still call the handler.

It may still be harmless, but we need to be sure of what is going on.

Reply:
> Because we might now be calling a handler that is now expected to be deleted.

Another thought I had was to use atomic.CompareAndSwap() to set isDeleted so we don't have to lock anything on informer deletion.  Then defer the actual removal from the handlers map to a goroutine that does lock the informer but it's from a goroutine so it doesn't matter.  Then when calling handlers, which we can still do under the informer lock, check isDeleted with CompareAndSwap() and skip if that is set.

But I wonder if there isn't a more general, more elegant solution to the issue of adding/removing informers that avoids the issue altogether.

Reply:
Does working on serializing all kubernetes events help? We may anyway have to do that for OVSDB bindings